src/algorithms/blocked_gibbs_algorithm.cc

```diff
@@ -14,9 +18,38 @@ void BlockedGibbsAlgorithm::print_startup_message() const {
 }
 
 void BlockedGibbsAlgorithm::sample_allocations() {
-  return;  // TODO
+  auto &rng = bayesmix::Rng::Instance().get();
+  unsigned int num_components = 0;  // TODO; maybe we should set init_num_cl...
+  for (int i = 0; i < data.rows(); i++) {
+    Eigen::VectorXd old_weights =
+        cond_mixing->get_weights(true, false, mix_covariates.row(i));
+    // Compute weights
+    Eigen::VectorXd logprobas(num_components);
+    for (int j = 0; j < num_components; j++) {
+      logprobas(j) = old_weights(j) + unique_values[j]->like_lpdf(
+                                          data.row(j), hier_covariates.row(j));
+    }
+    // Draw a NEW value for datum allocation
+    unsigned int c_new =
+        bayesmix::categorical_rng(stan::math::softmax(logprobas), rng, 0);
+    unsigned int c_old = allocations[i];
+    allocations[i] = c_new;
+    // Remove datum from old cluster, add to new
+    unique_values[c_old]->remove_datum(
+        i, data.row(i), update_hierarchy_params(), hier_covariates.row(i));
+    unique_values[c_new]->add_datum(i, data.row(i), update_hierarchy_params(),
+                                    hier_covariates.row(i));
+  }
 }
 
 void BlockedGibbsAlgorithm::sample_unique_values() {
-  return;  // TODO
+  for (auto &un : unique_values) {
+    // TODO should we only do it for nonempty clusters? if so, also in Neal2+
```

Comment:
we should do it for all clusters / components, but if there is no observation allocated to a cluster, the new values should be sampled from the prior (it might be that there is already such a logic inside the BaseHierarchy, but please double check)

Reply:
There is already such a logic indeed :) It's enforced by hand though: this function calls `get_posterior_parameters()` which in each one of our current hierarchies starts with
```
if (card == 0) {  // no update possible
  return *hypers;
}
...
```
It might be a good idea to enforce this automatically, for instance by adding a check before calling `get_posterior_parameters()`.

Reply:
That's true, I think this could be done at the `BaseHierarchy` level (i.e. do the check, if no data *sample from the prior*, else call the derived class method)

Reply:
Sure. One last thing: should we leave the checks in the individual `get_posterior_parameters()` anyway, just to be safe with future changes / in case someone calls it without actual data? The alternative is to assume that the given matrices are valid. I'm fine either way, but in the latter case we will make it explicitly clear in the documentation.

Reply:
I think those checks are fine, since the methods can be called also for different purposes and if there is no data the right value to return is the prior parameters